core/domain/feedback_services.py

```diff
@@ -406,16 +442,77 @@ def clear_feedback_message_references(user_id, exploration_id, thread_id):
-    feedback_message_reference = feedback_domain.FeedbackMessageReference(
-        exploration_id, thread_id, message_id)
 
-    for owner_id in exploration_rights.owner_ids:
-        owner_preferences = user_services.get_email_preferences(owner_id)
-        if (owner_id != author_id and
-                owner_preferences['can_receive_feedback_message_email']):
+    owner_ids = set(exploration_rights.owner_ids)
+    participant_ids = get_all_thread_participants(exploration_id, thread_id)
+    sender_id = set([author_id])
+
+    batch_recipient_ids = owner_ids - sender_id
+    other_recipient_ids = participant_ids - batch_recipient_ids - sender_id
+
+    return (batch_recipient_ids, other_recipient_ids)
+
+
+def send_batch_emails(recipient_list, feedback_message_reference):
+    """send feedback message email in batch."""
+    for recipient_id in recipient_list:
+        recipient_preferences = (
+            user_services.get_email_preferences(recipient_id))
+        if recipient_preferences['can_receive_feedback_message_email']:
             transaction_services.run_in_transaction(
-                _add_feedback_message_reference, owner_id,
+                _add_feedback_message_reference, recipient_id,
                 feedback_message_reference)
+
+
+def send_instant_emails(
+        recipient_list, feedback_message_reference, status_changed=False,
+        old_status=None, new_status=None):
+    """send feedback message email instantly."""
+    for recipient_id in recipient_list:
+        recipient_preferences = (
+            user_services.get_email_preferences(recipient_id))
+        if recipient_preferences['can_receive_feedback_message_email']:
+            transaction_services.run_in_transaction(
+                enqueue_feedback_message_instant_email_task, recipient_id,
+                feedback_message_reference, status_changed,
+                old_status, new_status)
+
+
+def add_message_to_email_buffer(
+        author_id, exploration_id, thread_id, message_id, message_length,
+        status_changed, old_status, new_status):
+    """sends email of feedback message to all thread participants.
+
+    sends email for new feedback message and change in status of thread.
+
+    Args:
+        author_id: id of author of message.
+        exploration_id: id of exploration that received new message.
+        thread_id: id of thread that received new message.
+        message_id: id of new message.
+        status_changed: True if thread status is changed else False.
+        old_status: value of old thread status.
+        new_status: value of new thread status.
+    """
+    feedback_message_reference = feedback_domain.FeedbackMessageReference(
+        exploration_id, thread_id, message_id)
+    batch_recipient_ids, other_recipient_ids = (
+        get_all_recipient_ids(exploration_id, thread_id, author_id))
+
+    if status_changed:
+        # send email for status change.
+        send_instant_emails(
```

Comment:
It seems rather convoluted to pass the status_changed conditional all over the place like this -- there ends up being a lot of "if status has changed" constructs all over the place.

I'm wondering if perhaps we should have two handlers in the feedback controller layer, one for status change emails and one for feedback message emails -- since they are sufficiently different things. Then we could just have something like:

```
    if old_status != new_status:
        _send_status_change_emails(other_recipient_ids, feedback_message_reference, old_status, new_status)
```